cmd/rocsp-tool/main.go

```diff
@@ -118,31 +154,285 @@ func main2() error {
+	ocspGenerator capb.OCSPGeneratorClient
+	clk           clock.Clock
+}
+
+func (cl *client) loadFromDB(ctx context.Context, speed ProcessingSpeed) error {
+	// To scan the DB efficiently, we want to select only currently-valid certificates. There's a
+	// handy expires index, but for selecting a large set of rows, using the primary key will be
+	// more efficient. So first we find a good id to start with, then scan from there. Note: since
+	// AUTO_INCREMENT can skip around a bit, we add padding to ensure we get all currently-valid
+	// certificates.
+	startTime := cl.clk.Now().Add(-24 * time.Hour)
+	var minID int64
+	err := cl.db.QueryRowContext(
+		ctx,
+		"SELECT MIN(id) FROM certificateStatus WHERE notAfter >= ?",
+		startTime,
+	).Scan(&minID)
+	if err != nil {
+		return fmt.Errorf("selecting minID: %w", err)
+	}
+
+	setDefault(&speed.RowsPerSecond, 2000)
+	setDefault(&speed.ParallelSigns, 100)
+
+	query := fmt.Sprintf("SELECT %s FROM certificateStatus WHERE id > ?",
+		strings.Join(sa.CertStatusMetadataFields(), ", "))
+	rows, err := cl.db.QueryContext(ctx, query, minID)
+	if err != nil {
+		return fmt.Errorf("scanning certificateStatus: %w", err)
+	}
+	defer rows.Close()
+
+	// Limit the rate of reading rows.
+	frequency := time.Duration(float64(time.Second) / float64(time.Duration(speed.RowsPerSecond)))
+	rowTicker := time.NewTicker(frequency)
+
+	statusesToSign := make(chan *sa.CertStatusMetadata)
+	successes := make(chan struct{}, speed.ParallelSigns)
+	errChan := make(chan error, speed.ParallelSigns)
+
+	var signersWg sync.WaitGroup
+
+	for i := 0; i < speed.ParallelSigns; i++ {
+		signersWg.Add(1)
+		go cl.signAndStoreResponses(statusesToSign, errChan, successes, &signersWg)
+	}
+	var n, errorCount int
+	// Consume all available successes and errors from the output channels.
+	// If none are immediately available, return.
+	consume := func() {
+		for {
+			select {
+			case <-successes:
+				n++
+			case err := <-errChan:
+				errorCount++
+				if errorCount < 10 {
+					log.Print(err)
+				} else if errorCount < 1000 && rand.Intn(1000) < 100 {
+					log.Print(err)
+				} else if errorCount < 100000 && rand.Intn(1000) < 10 {
+					log.Print(err)
+				} else if rand.Intn(1000) < 1 {
+					log.Print(err)
+				}
+			default:
+				return
+			}
+			if (n+errorCount)%10 == 0 {
+				log.Printf("stored %d response, %d errors", n, errorCount)
+			}
+		}
+	}
+	var scanned int
+	for rows.Next() {
+		<-rowTicker.C
+
+		status := new(sa.CertStatusMetadata)
+		if err := sa.ScanCertStatusMetadataRow(rows, status); err != nil {
+			return fmt.Errorf("scanning row %d: %w", n, err)
+		}
+		scanned++
+		if scanned%100000 == 0 {
+			log.Printf("scanned %d certificateStatus rows. current ID %d (note: subtract a safety margin from this if restarting)", scanned, status.ID)
+		}
+		statusesToSign <- status
+
+		consume()
+	}
+	rerr := rows.Close()
+	if rerr != nil {
+		return fmt.Errorf("closing rows: %w", rerr)
+	}
+
+	close(statusesToSign)
```

Comment:
Is there a reason this channel should be closed here as opposed to deferring the close after creating the channel?

Reply:
The `signAndStoreResponses` goroutines all keep looping until the `statusesToSign` channel is closed, so it has to be closed before the call to `signersWg.Wait()` below.

Reply:
The worker goroutines won't exit their `for` loop until the channel is closed, so if we did that `signersWg.Wait()` would block forever.